### ISSUE TITLE
MQTT publish subscribe in the same client

### ISSUE THREAD

**Issue by peguine** [eclipse/paho.mqtt.cpp#260]
hello everyone:

`mqtt::async_client mqtt_client(address, slaveId)`
when this client is connected , the main thread publishes messages and detach a thread to consume message
```
thread thListenMQTT(mqtt_listen,std::ref(mqtt_client));
thListenMQTT.detach();
```
```
void mqtt_listen(mqtt::async_client &client){

	while(true){
		if(client.is_connected()){
			client.start_consuming();
			client.subscribe(mqtt_sub_topic + slaveId,QOS)->wait();
			while(true){
				auto msg = client.consume_message();
				if(!msg) {
					cout << "mqtt consume error" <<endl;			
					break;
				}
				string response = Parse::ParseMsg (msg->get_payload_str());
				string topic = msg->get_topic();
				SendMQTT(topic,client,response);
			}
		}else{break;}
	}
	bMqttListenTh =false;
}
```

but when i pub to this client with mosquitto, 
`mqtt consume error`
MQTT error [-3]: Disconnected

so, what's problem and how to acheive this.

thanks guys!

**Comment by fpagliughi** [comment]
It sounds like you may be using the same Client ID for more than one client connection.

In MQTT, each client must connect to the broker with a unique Client ID. If you attempt a second connection with the same ID, it's up to the broker to decide what course of action to take. It could refuse to accept the second connection. Or it could accept the second connection and kick off (disconnect) the first. Most brokers seem to do the latter, and that makes sense. It prevents a stale connection from allowing you to reconnect.

Anyway, it's not widely advertised, but you can always connect to the broker using a blank client ID, in which case the broker will assign the client a temporary, unique ID.


So if you're using the same ID for the publisher and subscriber, then when you connect the publisher, the broker is disconnecting the subscriber.

**Comment by peguine** [comment]
> It sounds like you may be using the same Client ID for more than one client connection.
> 
> In MQTT, each client must connect to the broker with a unique Client ID. If you attempt a second connection with the same ID, it's up to the broker to decide what course of action to take. It could refuse to accept the second connection. Or it could accept the second connection and kick off (disconnect) the first. Most brokers seem to do the latter, and that makes sense. It prevents a stale connection from allowing you to reconnect.
> 
> Anyway, it's not widely advertised, but you can always connect to the broker using a blank client ID, in which case the broker will assign the client a temporary, unique ID.
> 
> So if you're using the same ID for the publisher and subscriber, then when you connect the publisher, the broker is disconnecting the subscriber.

I want to use the same client id for both the publisher and subscriber. and they need to process at the same time. because the subscriber is in a while loop, so i put it into a single thread. There is only one connection actually , the subscriber thread detaches when the connection is build .  Confused.

I'll try to subscribe in callback, rather a new thread. 

Thanks.

**Comment by peguine** [comment]
Unfortunately , using callback doesn't help . as soon as I publish message to the client by mosquitto ,the connection lost with a empty cause.
 so does mqtt  support publishing  and subscribing  with the same client.
here is what I am trying.

first， the callback is almost same with the sample "async subscribe"

```

class action_listener : public virtual mqtt::iaction_listener
{
	std::string name_;

	void on_failure(const mqtt::token& tok) override {
		std::cout << name_ << " failure";
		if (tok.get_message_id() != 0)
			std::cout << " for token: [" << tok.get_message_id() << "]" << std::endl;
		std::cout << std::endl;
	}

	void on_success(const mqtt::token& tok) override {
		std::cout << name_ << " success";
		if (tok.get_message_id() != 0)
			std::cout << " for token: [" << tok.get_message_id() << "]" << std::endl;
		auto top = tok.get_topics();
		if (top && !top->empty())
			std::cout << "\ttoken topic: '" << (*top)[0] << "', ..." << std::endl;
		std::cout << std::endl;
	}

public:
	action_listener(const std::string& name) : name_(name) {}
};

/////////////////////////////////////////////////////////////////////////////

/**
 * A callback class for use with the main MQTT client.
 */
class callback : public virtual mqtt::callback,
				public virtual mqtt::iaction_listener
{
	// Counter for the number of connection retries
	int nretry_;
	// The MQTT client
	mqtt::async_client& cli_;
	// Options to use if we need to reconnect
	mqtt::connect_options& connOpts_;
	// An action listener to display the result of actions.
	action_listener subListener_;
	string subTopic_;
public:
	void connection_lost(const string& cause) override {
		cout << "Connection lost" << endl;
		if (!cause.empty())
			cout << "\tcause: " << cause << endl;
	}

	void delivery_complete(mqtt::delivery_token_ptr tok) override {
		cout << "\tDelivery complete for token: "
			<< (tok ? tok->get_message_id() : -1) << endl;
	}

	// Re-connection failure
	void on_failure(const mqtt::token& tok) override {
		std::cout << "Connection attempt failed" << std::endl;
	}

	// (Re)connection success
	// Either this or connected() can be used for callbacks.
	void on_success(const mqtt::token& tok) override {}

	void connected(const std::string& cause) override {
		std::cout << "\nConnection success" << std::endl;
		std::cout << "\nSubscribing to topic '" << subTopic_ << "'\n"
			<< "\tfor client " << cli_.get_client_id()
			<< " using QoS" << QOS << "\n"<< std::endl;
		cli_.subscribe(subTopic_, QOS, nullptr, subListener_);
	}


	// Callback for when a message arrives.
	void message_arrived(mqtt::const_message_ptr msg) override {
		std::cout << "Message arrived" << std::endl;
		std::cout << "\ttopic: '" << msg->get_topic() << "'" << std::endl;
		std::cout << "\tpayload: '" << msg->to_string() << "'\n" << std::endl;
	}
public:
	callback(mqtt::async_client& cli, mqtt::connect_options& connOpts, string subTopic)
				: nretry_(0), cli_(cli), connOpts_(connOpts), subListener_("Subscriber"),subTopic_(subTopic) {}
};
```


```
	mqtt::connect_options connOpts;
	connOpts.set_keep_alive_interval(300);
	connOpts.set_clean_session(true);
	mqtt::async_client mqtt_client(address, slaveId);
	callback cb(mqtt_client,connOpts,mqtt_sub_topic + slaveId);
        ConnectMQTT(mqtt_client,cb,mqttconfig.Username,mqttconfig.Password);
        SendMQTT(topic,mqtt_client,data);
```


```
int ConnectMQTT(mqtt::async_client &client,callback &cb,std::string un="",std::string pwd=""){
	try {
		cout << "mqtt connecting ..." << endl;
		client.set_callback(cb);
	        mqtt::token_ptr conntok;
		mqtt::connect_options conopts(un,pwd);
		conntok= client.connect(conopts,nullptr,cb);
		cout << "mqtt waiting for the connection..." << endl;
		conntok->wait();
		cout << "mqtt connected clientid " << client.get_client_id()<< endl;

		return 0;
	}
	catch (const mqtt::exception& exc) {
		cerr <<"mqtt connect error: " << exc.what() << endl;
		return -1;
	}
}
```


when connected, SendMQTT publish successfully and the callback connected print right. so it's subscribing.

Once I publish something by mosquitto to this client, the connection lost.

**Comment by peguine** [comment]
thank you. i make a misitake in the mosquitto pub. so silly, sorry about that.

**Comment by fpagliughi** [comment]
No problem. We should do a better job of documenting this. It's come up a lot recently.
